docs/architecture/ADR-005-liquidation.md

```diff
@@ -0,0 +1,191 @@
+# ADR 005: Liquidation
+
+## Changelog
+
+- November 19, 2021: Initial Draft (@toteki)
+
+## Status
+
+Proposed
+
+## Context
+
+When borrowers on Umee exceed their borrow limit due to interest accrual or asset price fluctuations, their positions become eligible for liquidation by other users.
+
+Such users (Liquidators) pay off part of all of the borrower's loan, in exchange for a value of collateral equal to the amount paid off plus an additional percentage (the liquidation incentive).
+
+We must build the features necessary for liquidators to continuously look out for liquidation opportunities, then carry out chosen liquidations.
+
+Additional parameters will be required which define the liquidation incentive and any other new variables.
+
+## Decision
+
+Liquidation will require one message type (`MsgLiquidate`), one per-token parameter (`LiquidationIncentive`), and two global parameters (`MinimumCloseFactor` and `CompleteLiquidationThreshold`).
+
+There is no event type for when a borrower becomes a valid liquidation target, nor a list of valid targets stored in the module. Liquidators will have to use an off-chain tool to query their nodes periodically.
+
+## Detailed Design
+
+A function `IsLiquidationEligible(borrowerAddr)` can be created to determine if a borrower is currently exceeding their borrow limit. Any liquidation attempt against a borrower not over their limit should fail.
+
+A borrower's total borrower value (expressed in USD) can be computed from their total borrowed tokens and a price oracle.
+
+Their borrow limit is calculated similarly using borrower uToken balances, borrower collateral settings, current uToken exchange rates, and token collateral weights.
+
+### Basic Message Type
+
+To implement the liquidation functionality of the Asset Facility, one message type is required:
+
+```go
+// MsgLiquidate - a liquidator targets a specific borrower, asset type, and collateral type for complete or partial liquidation
+type MsgLiquidate struct {
+  Liquidator    sdk.AccAddress
+  Borrower      sdk.AccAddress
+  Repayment     sdk.Coin // denom + amount
+  RewardDenom   string
+}
+```
+
+Repayment's denom is the borrowed asset denom to be repaid (because the borrower may have multiple open borrows). It is always a base asset type (not a uToken).
+Its amount is the maximum amount of asset the liquidator is willing to repay. This field enables partial liquidation.
+
+RewardDenom is the collateral type which the liquidator will recieve in exchange for repaying the borrower's loan. It is always a uToken denomination.
```

Comment:
The liquidator can pick any token used as collateral in the protocol, right? is it possible for one to run out? Will there be races or highly preferable tokens chosen? If so, would that be an issue?

Reply:
Any token that the borrower themselves has lent and marked as collateral.

So if I'm a borrower and I've lent `ETH`, `uAtom`, and `uOsmo`, and enabled only `uAtom` and `uOsmo` as collateral, then a liquidator targeting my address has the option of receiving `u/uAtom` and `u/uOsmo` as a reward. Those rewards come directly out of my balance, so there is no highly contested system pool.

Reply:
Got it. Thanks!